Bug report: `npm install` fails at `exec:fixCryptoApiImports` on MacOS

**Describe the bug**
Can't run `npm i` to completion with current master on MacOS. Post-install job "exec:fixCryptoApiImports" fails with a `sed` error:

```
sed: can't read : No such file or directory
```

**To Reproduce**

I'm on MacOS 15.3.2, on a 2019 Intel MacBook Pro.

Installed `npm` using `nvm`, the problem occurs both with the latest LTS and with node@16 and npm@8 as [per the (possibly outdated) install instructions]().

Then simply ran `npm i`.

Output:

```
$ npm i
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@astronautlabs/amf@0.0.6',
npm WARN EBADENGINE   required: { node: '^14' },
npm WARN EBADENGINE   current: { node: 'v18.20.8', npm: '8.19.4' }
npm WARN EBADENGINE }
npm WARN deprecated osenv@0.1.5: This package is no longer supported.
npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm WARN deprecated rimraf@2.7.1: Rimraf versions prior to v4 are no longer supported
npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm WARN deprecated glob@8.1.0: Glob versions prior to v9 are no longer supported
npm WARN deprecated glob@7.1.7: Glob versions prior to v9 are no longer supported
npm WARN deprecated loglevel-message-prefix@3.0.0: Use @natlibfi/loglevel-message-prefix instead
npm WARN deprecated es6-polyfills@2.0.0: Use @natlibfi/es6-polyfills instead
npm WARN deprecated @babel/polyfill@7.12.1: 🚨 This package has been deprecated in favor of separate inclusion of a polyfill and regenerator-runtime (when needed). See the @babel/polyfill docs (https://babeljs.io/docs/en/babel-polyfill) for more information.
npm WARN deprecated bootstrap-colorpicker@3.4.0: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
npm WARN deprecated popper.js@1.16.1: You can find the new Popper v2 at @popperjs/core, this package is dedicated to the legacy v1
npm WARN deprecated core-js@2.6.12: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.
npm WARN deprecated core-js@2.6.12: core-js@<3.23.3 is no longer maintained and not recommended for usage due to the number of issues. Because of the V8 engine whims, feature detection in old core-js versions could cause a slowdown up to 100x even if nothing is polyfilled. Some versions have web compatibility issues. Please, upgrade your dependencies to the actual version of core-js.

> cyberchef@10.19.4 postinstall
> npx grunt exec:fixCryptoApiImports && npx grunt exec:fixSnackbarMarkup && npx grunt exec:fixJimpModule

Running "exec:fixCryptoApiImports" (exec) task
>> sed: can't read : No such file or directory
>> Exited with code: 1.
>> Error executing child process: Error: Process exited with code 1.
Warning: Task "exec:fixCryptoApiImports" failed. Use --force to continue.

Aborted due to warnings.
npm ERR! code 3
npm ERR! path /Users/bart/src/ext/CyberChef
npm ERR! command failed
npm ERR! command sh -c -- npx grunt exec:fixCryptoApiImports && npx grunt exec:fixSnackbarMarkup && npx grunt exec:fixJimpModule

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/bart/.npm/_logs/2025-04-03T10_55_46_184Z-debug-0.log
```


**Expected behaviour**

No errors during `npm i`.


**Desktop:**
 - OS: MacOS 15.3.2 
 - Browser: N/A
 - CyberChef version: 10.19.4 (current master)


**Additional context**

This seems to be related with the special case for "darwin" in the affected job. If I remove the `''` argument to the `-i` command line parameter to `sed`, the job completes. Presumably, at some point, MacOS started shipping a more compatible version of `sed`.

--- a/Gruntfile.js
+++ b/Gruntfile.js
@@ -4,9 +4,19 @@ const webpack = require("webpack");
 const HtmlWebpackPlugin = require("html-webpack-plugin");
 const BundleAnalyzerPlugin = require("webpack-bundle-analyzer").BundleAnalyzerPlugin;
 const glob = require("glob");
-const path = require("path");
+const path = require("node:path");
+const cp = require("node:child_process");
 
 const nodeFlags = "--no-warnings --no-deprecation";
+
+// Prepare platform-dependent commands
+
+// Older MacOS versions don't come with `sha256sum`, but they all come with `shasum`
+const sha256sumCmd = process.platform === "darwin" ? "shasum -a 256" : "sha256sum";
+
+// MacOS (and FreeBSD, but not OpenBSD) require an argument to `-i`.
+// However, users may have installed GNU sed, so let's check what `sed` says itself.
+const sedCmd = cp.execSync("sed -i 2>&1 | head -n1").toString().includes("option requires an argument") ? "sed -i ''" : "sed -i";
 
 /**
  * Grunt configuration for building the app in various formats.
@@ -60,7 +70,7 @@ module.exports = function (grunt) {
 
     grunt.registerTask("findModules",
         "Finds all generated modules and updates the entry point list for Webpack",
-        function(arg1, arg2) {
+        function (arg1, arg2) {
             const moduleEntryPoints = listEntryModules();
 
             grunt.log.writeln(`Found ${Object.keys(moduleEntryPoints).length} modules.`);
@@ -114,7 +124,7 @@ module.exports = function (grunt) {
                 output: {
                     path: __dirname + "/build/prod",
                     filename: chunkData => {
-                        return chunkData.chunk.name === "main" ? "assets/[name].js": "[name].js";
+                        return chunkData.chunk.name === "main" ? "assets/[name].js" : "[name].js";
                     },
                     globalObject: "this"
                 },
@@ -334,20 +344,10 @@ module.exports = function (grunt) {
         },
         exec: {
             calcDownloadHash: {
-                command: function () {
-                    switch (process.platform) {
-                        case "darwin":
-                            return chainCommands([
-                                `shasum -a 256 build/prod/${downloadZipFilename} | awk '{print $1;}' > build/prod/sha256digest.txt`,
-                                `sed -i '' -e "s/DOWNLOAD_HASH_PLACEHOLDER/$(cat build/prod/sha256digest.txt)/" build/prod/index.html`
-                            ]);
-                        default:
-                            return chainCommands([
-                                `sha256sum build/prod/${downloadZipFilename} | awk '{print $1;}' > build/prod/sha256digest.txt`,
-                                `sed -i -e "s/DOWNLOAD_HASH_PLACEHOLDER/$(cat build/prod/sha256digest.txt)/" build/prod/index.html`
-                            ]);
-                    }
-                },
+                command: chainCommands([
+                    `${sha256sumCmd} build/prod/${downloadZipFilename} | awk '{print $1;}' > build/prod/sha256digest.txt`,
+                    `${sedCmd} -e "s/DOWNLOAD_HASH_PLACEHOLDER/$(cat build/prod/sha256digest.txt)/" build/prod/index.html`,
+                ]),
             },
             repoSize: {
                 command: chainCommands([

--- a/Gruntfile.js
+++ b/Gruntfile.js
@@ -75,17 +75,23 @@ module.exports = function (grunt) {
         });
 
     grunt.registerTask("calcDownloadHash", "Compute download hash", function () {
-        const computeDigest = p => {
-            const content = fs.readFileSync(p);
-            const hash = crypto.hash("sha256", content);
-            return hash;
-        };
+        const done = this.async();
 
-        const digest = computeDigest(`build/prod/${downloadZipFilename}`);
-        fs.writeFileSync("build/prod/sha256digest.txt", `${digest}\n`, { encoding: "utf8" });
+        const hash = crypto.createHash("sha256");
 
-        const index = fs.readFileSync("build/prod/index.html", { encoding: "utf8" });
-        fs.writeFileSync("build/prod/index.html", index.replace(/DOWNLOAD_HASH_PLACEHOLDER/g, digest), { encoding: "utf8" });
+        // Use online algorithm to calculate hash, prevents reading the entire 75+ MB zip file into memory
+        fs.createReadStream(`build/prod/${downloadZipFilename}`)
+            .on("data", (chunk) => hash.update(chunk))
+            .on("end", () => {
+                const digest = hash.digest("hex");
+                fs.writeFileSync("build/prod/sha256digest.txt", `${digest}\n`, { encoding: "utf8" });
+
+                const index = fs.readFileSync("build/prod/index.html", { encoding: "utf8" });
+                fs.writeFileSync("build/prod/index.html", index.replace(/DOWNLOAD_HASH_PLACEHOLDER/g, digest), { encoding: "utf8" });
+
+                done(true);
+            })
+            .on("error", (err) => done(false));
     });
 
 

--- a/Gruntfile.js
+++ b/Gruntfile.js
@@ -1,22 +1,15 @@
 "use strict";
 
+const crypto = require("node:crypto");
+const fs = require("node:fs");
+const path = require("node:path");
+
 const webpack = require("webpack");
 const HtmlWebpackPlugin = require("html-webpack-plugin");
 const BundleAnalyzerPlugin = require("webpack-bundle-analyzer").BundleAnalyzerPlugin;
 const glob = require("glob");
-const path = require("node:path");
-const cp = require("node:child_process");
 
 const nodeFlags = "--no-warnings --no-deprecation";
-
-// Prepare platform-dependent commands
-
-// Older MacOS versions don't come with `sha256sum`, but they all come with `shasum`
-const sha256sumCmd = process.platform === "darwin" ? "shasum -a 256" : "sha256sum";
-
-// MacOS (and FreeBSD, but not OpenBSD) require an argument to `-i`.
-// However, users may have installed GNU sed, so let's check what `sed` says itself.
-const sedCmd = cp.execSync("sed -i 2>&1 | head -n1").toString().includes("option requires an argument") ? "sed -i ''" : "sed -i";
 
 /**
  * Grunt configuration for building the app in various formats.
@@ -39,7 +32,7 @@ module.exports = function (grunt) {
         "Creates a production-ready build. Use the --msg flag to add a compile message.",
         [
             "eslint", "clean:prod", "clean:config", "exec:generateConfig", "findModules", "webpack:web",
-            "copy:standalone", "zip:standalone", "clean:standalone", "exec:calcDownloadHash", "chmod"
+            "copy:standalone", "zip:standalone", "clean:standalone", "calcDownloadHash", "chmod"
         ]);
 
     grunt.registerTask("node",
@@ -80,6 +73,20 @@ module.exports = function (grunt) {
                     main: "./src/web/index.js"
                 }, moduleEntryPoints));
         });
+
+    grunt.registerTask("calcDownloadHash", "Compute download hash", function () {
+        const computeDigest = p => {
+            const content = fs.readFileSync(p);
+            const hash = crypto.hash("sha256", content);
+            return hash;
+        };
+
+        const digest = computeDigest(`build/prod/${downloadZipFilename}`);
+        fs.writeFileSync("build/prod/sha256digest.txt", `${digest}\n`, { encoding: "utf8" });
+
+        const index = fs.readFileSync("build/prod/index.html", { encoding: "utf8" });
+        fs.writeFileSync("build/prod/index.html", index.replace(/DOWNLOAD_HASH_PLACEHOLDER/g, digest), { encoding: "utf8" });
+    });
 
 
     // Load tasks provided by each plugin
@@ -343,12 +350,6 @@ module.exports = function (grunt) {
             }
         },
         exec: {
-            calcDownloadHash: {
-                command: chainCommands([
-                    `${sha256sumCmd} build/prod/${downloadZipFilename} | awk '{print $1;}' > build/prod/sha256digest.txt`,
-                    `${sedCmd} -e "s/DOWNLOAD_HASH_PLACEHOLDER/$(cat build/prod/sha256digest.txt)/" build/prod/index.html`,
-                ]),
-            },
             repoSize: {
                 command: chainCommands([
                     "git ls-files | wc -l | xargs printf '\n%b\ttracked files\n'",

--- a/package.json
+++ b/package.json
@@ -63,6 +63,7 @@
     "css-loader": "^7.1.4",
     "eslint": "^9.39.5",
     "eslint-plugin-jsdoc": "^50.8.0",
+    "glob": "^13.0.6",
     "globals": "^17.7.0",
     "grunt": "^1.6.2",
     "grunt-chmod": "~1.1.1",

